packages/contracts/contracts/zkopru/controllers/UserInteractable.sol

```diff
@@ -178,17 +181,19 @@ contract UserInteractable is Storage {
         uint256[3] memory resultHashInputs = [spendingPubKey, salt, assetHash];
         uint256 note = Poseidon3.poseidon(resultHashInputs);
         // Receive token
-        if (token != address(0) && amount != 0) {
-            try
-                IERC20(token).transferFrom(msg.sender, address(this), amount)
-            {} catch {
-                revert("Transfer ERC20 failed");
-            }
-        } else if (token != address(0)) {
-            try
-                IERC721(token).transferFrom(msg.sender, address(this), nft)
-            {} catch {
-                revert("Transfer NFT failed");
+        if (token != address(0)) {
+            if (chain.registeredERC20s[token]) {
+                require(nft == 0, "ERC20 note cannot have NFT");
+                IERC20(token).safeTransferFrom(
+                    msg.sender,
+                    address(this),
+                    amount
+                );
+            } else if (chain.registeredERC721s[token]) {
+                require(amount == 0, "NFT note cannot have amount");
+                IERC721(token).transferFrom(msg.sender, address(this), nft);
```

Comment:
Consider using `safeTransferFrom` unless there's a reason to allow transfers to the 0 address.

Reply:
resolve by https://github.com/zkopru-network/zkopru/pull/206/commits/4a125758c92afe09626df3548266b8a81df132aa and https://github.com/zkopru-network/zkopru/pull/206/commits/c95b341b173bd60e24d417a5b4423ae284edfb02